# 196. IMPLEMENTANDO NUESTRO PROPIO ACP EN PYTHON 

## ANALISIS DE COMPONENTES PRINCIPALES - Paso a Paso 

Problema de las dimensiones, el objetivo es identificar patrones, correlaciones entre variables, de modo que variables que esten muy correlacionadas, las intentemos juntar en una nueva, para ir quitando dimension. Hacer esto solo cuando deba de ser oportuno. El analisis de componentes principales. 
## Tecnica del analisis del discriminate lineal - Transformacion lineal 
La ACP busca las componentes principales que maximizan la varianza del dato, el ADL intenta buscar direcciones que maximizen la separacion, la discriminacion entre las distintas clases, lo cual puede ser util, cuando se debe de aplicar algoritmos de clasificacion de patrones, etc. 

Objetivo reducir de un espacio n - dimension en p - dimensional. 
Para reducir el coste computacional del algoritmo que apliquemos a posteriori. 
- ¿Que tan pequeño podra ser el ACP?

* Estandarizar los datos 
* Obtener los vectores y valores propios a partir de la matriz de covarianzas o de correlaciones o incluso la tecninca de singura vector de descomposicion. 
* Ordenar los valores propios en orden descendente y quedarnos con los *p* que se correspondan a los *p* mayores y asi disminuir el numero de variables del dataset. 
* Contruir la matriz de proyeccion W a apartir de los p vectores propios
* Transformar del dataset original X a traves de W para asi obtener datos en el subespacio dimensional de dimension *p*, que sera Y. 

In [36]:
import pandas as pd 

In [37]:
df = pd.read_csv("C:/Users/URIEL/Downloads/python-ml-course-master (2)/datasets/iris/iris.csv")

In [38]:
df.head()

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width Species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

# 198. PLOTLY, LA LIBRERIA DE GRAFICOS PERSONALIZADOS E INTERACTIVOS

In [39]:
#Divideremos el data set en X e Y, siendo X el dataset con el que trabajaremos. 
X = df.iloc[:,0:4].values
y = df.iloc [:,4].values

In [40]:
X #CADA FILA ES UN VECTOR DE UN ESPACIO VECTORIAL DE DIMENSION 4. 
#CADA COLUMNA TIENE UN SENTIDO. 


array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

In [41]:
#Haremos un histograma 
import chart_studio.plotly as py
from plotly.graph_objs import *
import plotly.tools as tls
#herramientas de dibujo. 
#haremos histogramas distintos, para cada una de las variables, dentro del mismo histograma. 

In [42]:
import chart_studio.plotly as py

import plotly.graph_objects as ir

In [43]:
#Haremos histogramas por cada una de las variables, todas dentro del mismo histograma.
traces = [] #array vacio 

legend = {0:False, 1:False, 2:False, 3:False} #diccionario para las 4 variables. 


#Estos son los colores para cada flor. 
colors = {'setosa': 'rgb(255,127,20)', #elegimos el color de tal manera que predomina un color para cada una de ellas. 

         'versicolor': 'rgb(31, 220, 120)',

         'virginica': 'rgb(44, 50, 180)'} #que tenga mucho azul. 


#construir la coloracion, una para cada una de las dimensiones del problema. 
for col in range(4): 

    for key in colors:

        traces.append(ir.Histogram(x=X[y==key, col], opacity = 0.7,

                                xaxis="x%s"%(col+1), marker=ir.Marker(color=colors[key]),

                               name = key, showlegend=legend[col]))

    legend = {0:False, 1:False, 2:False, 3:False}

       

data = ir.Data(traces)

layout = ir.Layout(barmode="overlay",

                xaxis=ir.XAxis(domain=[0,0.25], title="Long. Sépalos (cm)"),

                xaxis2=ir.XAxis(domain=[0.3, 0.5], title = "Anch. Sépalos (cm)"),

                xaxis3=ir.XAxis(domain = [0.55, 0.75], title = "Long. Pétalos (cm)"),

                xaxis4=ir.XAxis(domain=[0.8,1.0], title = "Anch. Pétalos (cm)"),

                yaxis=ir.YAxis(title="Número de ejemplares"),

                title="Distribución de los rasgos de las diferentes flores Iris")



fig = ir.Figure(data = data, layout = layout)

fig.show()

C:\Users\URIEL\anaconda3\lib\site-packages\plotly\graph_objs\_deprecations.py:434: DeprecationWarning:

plotly.graph_objs.Marker is deprecated.
Please replace it with one of the following more specific types
  - plotly.graph_objs.scatter.Marker
  - plotly.graph_objs.histogram.selected.Marker
  - etc.


C:\Users\URIEL\anaconda3\lib\site-packages\plotly\graph_objs\_deprecations.py:31: DeprecationWarning:

plotly.graph_objs.Data is deprecated.
Please replace it with a list or tuple of instances of the following types
  - plotly.graph_objs.Scatter
  - plotly.graph_objs.Bar
  - plotly.graph_objs.Area
  - plotly.graph_objs.Histogram
  - etc.


C:\Users\URIEL\anaconda3\lib\site-packages\plotly\graph_objs\_deprecations.py:544: DeprecationWarning:

plotly.graph_objs.XAxis is deprecated.
Please replace it with one of the following more specific types
  - plotly.graph_objs.layout.XAxis
  - plotly.graph_objs.layout.scene.XAxis


C:\Users\URIEL\anaconda3\lib\site-packages\plotly\graph_objs\_depreca

In [44]:
#para estandarizar 
from sklearn.preprocessing import StandardScaler

In [13]:
X_std = StandardScaler().fit_transform(X)
X_std

array([[-9.00681170e-01,  1.01900435e+00, -1.34022653e+00,
        -1.31544430e+00],
       [-1.14301691e+00, -1.31979479e-01, -1.34022653e+00,
        -1.31544430e+00],
       [-1.38535265e+00,  3.28414053e-01, -1.39706395e+00,
        -1.31544430e+00],
       [-1.50652052e+00,  9.82172869e-02, -1.28338910e+00,
        -1.31544430e+00],
       [-1.02184904e+00,  1.24920112e+00, -1.34022653e+00,
        -1.31544430e+00],
       [-5.37177559e-01,  1.93979142e+00, -1.16971425e+00,
        -1.05217993e+00],
       [-1.50652052e+00,  7.88807586e-01, -1.34022653e+00,
        -1.18381211e+00],
       [-1.02184904e+00,  7.88807586e-01, -1.28338910e+00,
        -1.31544430e+00],
       [-1.74885626e+00, -3.62176246e-01, -1.34022653e+00,
        -1.31544430e+00],
       [-1.14301691e+00,  9.82172869e-02, -1.28338910e+00,
        -1.44707648e+00],
       [-5.37177559e-01,  1.47939788e+00, -1.28338910e+00,
        -1.31544430e+00],
       [-1.26418478e+00,  7.88807586e-01, -1.22655167e+00,
      

In [14]:
traces = []

legend = {0:False, 1:False, 2:False, 3:False}



colors = {'setosa': 'rgb(255,127,20)',

         'versicolor': 'rgb(31, 220, 120)',

         'virginica': 'rgb(44, 50, 180)'}



for col in range(4):

    for key in colors:

        traces.append(ir.Histogram(x=X_std[y==key, col], opacity = 0.7,

                                xaxis="x%s"%(col+1), marker=ir.Marker(color=colors[key]),

                               name = key, showlegend=legend[col]))

    legend = {0:False, 1:False, 2:False, 3:False}

       

data = ir.Data(traces)

layout = ir.Layout(barmode="overlay",

                xaxis=ir.XAxis(domain=[0,0.25], title="Long. Sépalos (cm)"),

                xaxis2=ir.XAxis(domain=[0.3, 0.5], title = "Anch. Sépalos (cm)"),

                xaxis3=ir.XAxis(domain = [0.55, 0.75], title = "Long. Pétalos (cm)"),

                xaxis4=ir.XAxis(domain=[0.8,1.0], title = "Anch. Pétalos (cm)"),

                yaxis=ir.YAxis(title="Número de ejemplares"),

                title="Distribución de los rasgos de las diferentes flores Iris")



fig = ir.Figure(data = data, layout = layout)

fig.show()
#todos los datos centrados en 0. 

C:\Users\URIEL\anaconda3\lib\site-packages\plotly\graph_objs\_deprecations.py:434: DeprecationWarning:

plotly.graph_objs.Marker is deprecated.
Please replace it with one of the following more specific types
  - plotly.graph_objs.scatter.Marker
  - plotly.graph_objs.histogram.selected.Marker
  - etc.


C:\Users\URIEL\anaconda3\lib\site-packages\plotly\graph_objs\_deprecations.py:31: DeprecationWarning:

plotly.graph_objs.Data is deprecated.
Please replace it with a list or tuple of instances of the following types
  - plotly.graph_objs.Scatter
  - plotly.graph_objs.Bar
  - plotly.graph_objs.Area
  - plotly.graph_objs.Histogram
  - etc.


C:\Users\URIEL\anaconda3\lib\site-packages\plotly\graph_objs\_deprecations.py:544: DeprecationWarning:

plotly.graph_objs.XAxis is deprecated.
Please replace it with one of the following more specific types
  - plotly.graph_objs.layout.XAxis
  - plotly.graph_objs.layout.scene.XAxis


C:\Users\URIEL\anaconda3\lib\site-packages\plotly\graph_objs\_depreca

# 199. LOS VALORES Y VECTORES PROPIOS DE LA MATRIZ DE COVARIANZAS.

Despues de haber hecho la visualizacion de los datos y la estandarizacion, procederemos con la primera de las etapas. 

### 1 CALCULAR LA DESCOMPOSICION DE VALORES Y VECTORES PROPIOS. 
Un vector y valor  propio representan el core fundamental del ACP, apartir de la matriz de covarianzas se extraen los vectores propios, las componentes principales, que determinan las direcciones en el nuevo espacio vectorial, mientras los valores propios determinan la magnitud. 

In [15]:
from IPython.display import display, Math, Latex 

In [16]:
display(Math(r'\sigma_{jk} = \frac{1}{n-1}\sum_{i=1}^m (x_{ij} - \overline{x_j}) (x_{ik} - \overline{x_k})'))
#Esta es la entrada jk de la matriz de covarianzas. 

<IPython.core.display.Math object>

In [17]:
#para resumirla en una formula unica, formula de la matriz de covarianzas.  
display(Math(r'\Sigma = \frac{1}{n-1}((X-\overline{x})^T(X - \overline{x}))'))

<IPython.core.display.Math object>

In [18]:
import numpy as np 

In [19]:
mean_vect = np.mean(X_std, axis = 0) #por filas, la media de las 4 variables. 
mean_vect #vector de medias. 

array([-4.73695157e-16, -7.81597009e-16, -4.26325641e-16, -4.73695157e-16])

In [20]:
#creamos la matriz de covarianzas. 
cov_matrix=  (X_std - mean_vect).T.dot((X_std - mean_vect)) /(X_std.shape[0]-1)
print("La matriz de covarianzas es:")
cov_matrix


La matriz de covarianzas es:


array([[ 1.00671141, -0.11835884,  0.87760447,  0.82343066],
       [-0.11835884,  1.00671141, -0.43131554, -0.36858315],
       [ 0.87760447, -0.43131554,  1.00671141,  0.96932762],
       [ 0.82343066, -0.36858315,  0.96932762,  1.00671141]])

haremos la descomposicion de la matriz de covarianzas, para ello, una ves que tenemos la matriz, para calcular los valores y vectores 
propios se tendria que resolver el sistema de ecuaciones, hacer el determinamente, etc.
## ESTO SERIA USANDO LA MATRIZ DE COVARIANZAS
Porque existen otros modos de hacer esta descomposicion.

In [21]:
eig_vals, eig_vectors = np.linalg.eig(cov_matrix)
print("Valores propios \n%s"%eig_vals)
print("Vectores propios \n%s"%eig_vectors)

Valores propios 
[2.93808505 0.9201649  0.14774182 0.02085386]
Vectores propios 
[[ 0.52106591 -0.37741762 -0.71956635  0.26128628]
 [-0.26934744 -0.92329566  0.24438178 -0.12350962]
 [ 0.5804131  -0.02449161  0.14212637 -0.80144925]
 [ 0.56485654 -0.06694199  0.63427274  0.52359713]]


# 200. LA MATRIZ DE CORRELACIONES Y EL SINGULAR VALUE DESCOMPOSITION.

VERSION 2. 
- Descomposicion usando la matriz de correlaciones. 
* esto es muy tipico en el campo de las finanzas, la matriz de correlaciones es una versiona normalizada de la matriz de covarianzas. 

In [22]:
corr_matrix = np.corrcoef(X_std.T)
corr_matrix

array([[ 1.        , -0.11756978,  0.87175378,  0.81794113],
       [-0.11756978,  1.        , -0.4284401 , -0.36612593],
       [ 0.87175378, -0.4284401 ,  1.        ,  0.96286543],
       [ 0.81794113, -0.36612593,  0.96286543,  1.        ]])

In [23]:
eig_vals_corr, eig_vectors_corr =  np.linalg.eig(corr_matrix)
print("Valores propios \n%s"%eig_vals_corr)
print("Vectores propios \n%s"%eig_vectors_corr)
#la matriz cambia un poco. 

Valores propios 
[2.91849782 0.91403047 0.14675688 0.02071484]
Vectores propios 
[[ 0.52106591 -0.37741762 -0.71956635  0.26128628]
 [-0.26934744 -0.92329566  0.24438178 -0.12350962]
 [ 0.5804131  -0.02449161  0.14212637 -0.80144925]
 [ 0.56485654 -0.06694199  0.63427274  0.52359713]]


### SINGULAR VALUE DECOMPOSITION 

Cuando se debe de llevar acabo un ACP digno, las implementaciones que aparecen en python, no lo hacen asi, sino que hacen esta tercara tecnica, ya que mejora mucho la eficacia computacional. 

In [24]:
u, s, v = np.linalg.svd(X_std.T)
u #da la misma matriz de vectores propios. 

array([[-0.52106591, -0.37741762,  0.71956635,  0.26128628],
       [ 0.26934744, -0.92329566, -0.24438178, -0.12350962],
       [-0.5804131 , -0.02449161, -0.14212637, -0.80144925],
       [-0.56485654, -0.06694199, -0.63427274,  0.52359713]])

In [25]:
s #este es mucho mas grande que los otros valores propios. 

array([20.92306556, 11.7091661 ,  4.69185798,  1.76273239])

In [26]:
v

array([[ 1.08239531e-01,  9.94577561e-02,  1.12996303e-01, ...,
        -7.27030413e-02, -6.56112167e-02, -4.59137323e-02],
       [-4.09957970e-02,  5.75731483e-02,  2.92000319e-02, ...,
        -2.29793601e-02, -8.63643414e-02,  2.07800179e-03],
       [ 2.72186462e-02,  5.00034005e-02, -9.42089147e-03, ...,
        -3.84023516e-02, -1.98939364e-01, -1.12588405e-01],
       ...,
       [ 5.43380310e-02,  5.12936114e-03,  2.75184277e-02, ...,
         9.89532683e-01, -1.41206665e-02, -8.30595907e-04],
       [ 1.96438400e-03,  8.48544595e-02,  1.78604309e-01, ...,
        -1.25488246e-02,  9.52049996e-01, -2.19201906e-02],
       [ 2.46978090e-03,  5.83496936e-03,  1.49419118e-01, ...,
        -7.17729676e-04, -2.32048811e-02,  9.77300244e-01]])

# 201. LA SELECCION DE LAS COMPONENTES PRINCIPALES

Hora de seleccionar las componentes principales, el objetivo del ACP, es reducir la dimensionalidad del espacio vectorial original, proyectando los datos en subespacios de dimension mas pequeña desde las direcciones que definen los vectores propios desde los ejes, sin embargo los vectores propios no solo definen la direccion sino que ademas deben tener dimension uno, para que se trate de una base de un subespacio vectorial. para que la tecnica funciones todos deben tener longitud uno.

In [27]:
#para ver que tengan longitud uno. 
for ev in eig_vectors:
    print("La longitud del VP es : %s"%np.linalg.norm(ev))
#de todos los vectores propios que nos podia devolver el metodo, no solo ha devuelto un vector propio para cada uno de los vectores propios
#sino que todos tienen longitud uno. 

La longitud del VP es : 0.9999999999999999
La longitud del VP es : 1.0
La longitud del VP es : 0.9999999999999999
La longitud del VP es : 1.0


para decidir que vectores propios se pueden eliminar sin perder tanta informacion para construir un espacio dimensional de dimension 
inferior, buscaremos cada uno de los valores propios. 
* Los vectores propios que tengan el valor propio con menos informacion de la distrbucion de los datos, seran eliminados, para ello se hara un ranking de los mas altas hasta el mas bajo. 


In [28]:
eigen_pairs = [ (np.abs(eig_vals[i]), eig_vectors[:,i]) for i in range(len(eig_vals))]
eigen_pairs
#ordenamos los vectores propios con valor propio del mayor a menor. 

[(2.938085050199995,
  array([ 0.52106591, -0.26934744,  0.5804131 ,  0.56485654])),
 (0.9201649041624871,
  array([-0.37741762, -0.92329566, -0.02449161, -0.06694199])),
 (0.14774182104494812,
  array([-0.71956635,  0.24438178,  0.14212637,  0.63427274])),
 (0.020853862176462873,
  array([ 0.26128628, -0.12350962, -0.80144925,  0.52359713]))]

In [29]:
#si los queremos ordenar. 
eigen_pairs.sort()
eigen_pairs
#valor propio del mas alto al mas pequeño. 

[(0.020853862176462873,
  array([ 0.26128628, -0.12350962, -0.80144925,  0.52359713])),
 (0.14774182104494812,
  array([-0.71956635,  0.24438178,  0.14212637,  0.63427274])),
 (0.9201649041624871,
  array([-0.37741762, -0.92329566, -0.02449161, -0.06694199])),
 (2.938085050199995,
  array([ 0.52106591, -0.26934744,  0.5804131 ,  0.56485654]))]

In [30]:
print("Valores propios en orden descendente:")
for ep in eigen_pairs:
    print(ep[0])

Valores propios en orden descendente:
0.020853862176462873
0.14774182104494812
0.9201649041624871
2.938085050199995


¿Cuantas componentes principales debemos elegir para el nuevo subespacio?, esta claro que en el viejo teniamos 4, pero nos podemos quedar con 3, 2 o 1. Una medida muy util es la llamada explicacion de la varianza, que se puede calcular apartir de los valores propios. La varianza explicativa nos dice cuanta informacion se puede atribuir a cada una de las componentes principales y luego se puede llevar acabo un grafico acumulativo para ver si solo me quedare con 1, 2 o con todas. 
Para ello: 

In [31]:
total_sum = sum(eig_vals) 
#la explicacion de la varianza sera: 
var_exp = [(i/total_sum)*100 for i in sorted(eig_vals, reverse = True)]

In [32]:
var_exp
#lo que nos dice esto, es que quedandonos solo con una dimension, explicamos el 72% de la variabilidad, el segundo explica un 22% adicional 


[72.96244541329987, 22.85076178670176, 3.6689218892828777, 0.5178709107154946]

In [33]:
cum_var_exp = np.cumsum(var_exp)
cum_var_exp

array([ 72.96244541,  95.8132072 ,  99.48212909, 100.        ])

In [34]:
import chart_studio.plotly as py

import plotly.graph_objects as ir

In [35]:
#haremos un grafico que combine varias formas, usando plotly. 
plot1 = ir.Bar(x=["CP %s"%i for i in range(1,5)], y = var_exp, showlegend=False)
plot2 = ir.Scatter(x=["CP %s"%i for i in range(1,5)], y=cum_var_exp, showlegend=True, name = "% de Varianza Explicada Acumulada")

data = ir.Data([plot1, plot2])

layout = ir.Layout(xaxis = XAxis(title = "Componentes principales"), 
               yaxis = YAxis(title = "Porcentaje de varianza explicada"), 
               title = "Porcentaje de variabilidad explicada por cada componente principal")

fig = ir.Figure(data = data, layout = layout)
fig.show()

* Podemos ver las 4 componentes principales, si solo nos quedaramos con una componente principal, se puede ver que explicaria el 72.96% de la varianza.
* Si nos quedamos con la 1ra y 2da nos quedamos con el 95.81%. 

Ahora construimos la matriz de proyección, que es concatenar los p vectores propios mas grandes que se quieran conservar de datos. reduciremos la dimension del espacio vectorial, elegiendo los 2 vectores propios con mayor valor propio. 

In [79]:
W = np.hstack((eigen_pairs[0][1].reshape(4,1), 
              eigen_pairs[1][1].reshape(4,1)))
W #matriz de 4x2 que ya esta preparada para poder hacer la proyeccion. 


array([[ 0.26128628, -0.71956635],
       [-0.12350962,  0.24438178],
       [-0.80144925,  0.14212637],
       [ 0.52359713,  0.63427274]])

Lo que haremos ahora es proyectar los datos al nuevo espacio vectorial y asi poder trabajr con nuevas dimensiones de las cuales habiamos partido. 

# 202. LA PROYECCION EN EL SUBESPACIO VECTORIAL RESULTANTE

### PROYECTANDO LAS VARIABLES EN EL NUEVO SUBESPACIO VECTORIAL

In [82]:
#reduciremos a un espacio de 4 a 2 dimensiones. 
display(Math(r'Y = X \cdot W, X \in M(\mathbb R)_{150,4}, W \in M(\mathbb R)_{4,2}, Y \in M(\mathbb R)_{150,2}'))

<IPython.core.display.Math object>

In [83]:
Y = X_std.dot(W)

In [92]:
#haremos la representacion grafico de la representacion. 
results = []
for name in ('setosa', 'versicolor', 'virginica'):
    result = ir.Scatter(x=Y[y==name,0], y=Y[y==name, 1], 
                    mode = "markers", name=name, marker=ir.Marker(size=12, line = Line(color='rgba(220,220,220,0.15)', 
                                                                                   width=0.5),opacity=0.8))
    results.append(result)
    
data = ir.Data(results)
layout = ir.Layout(showlegend=True, scene = Scene(xaxis = XAxis(title = "Componente principal 1"), 
                                                 yaxis = YAxis(title = "Componente principal 2")))

fig = ir.Figure(data=data, layout=layout)
fig.show()

Ya que hemos reducido la dimensionalidad de los datos, podemos hacer algun algoritmo de interes, como una clasificacion, una regresion, machine learning con alguna red neuronal o con alguna otra tecnica. 

# 203. IMPLEMENTACION DE ACP CON SKLEARN 

In [13]:
#usamos todos los mismos datos anteriores, al igual que las mismas librerias. 
from sklearn.decomposition import PCA as sk_pca

In [14]:
acp = sk_pca(n_components = 2)
Y = acp.fit_transform(X_std)

In [15]:
Y #Esta seria la transformacion de los datos. 

array([[-2.26470281,  0.4800266 ],
       [-2.08096115, -0.67413356],
       [-2.36422905, -0.34190802],
       [-2.29938422, -0.59739451],
       [-2.38984217,  0.64683538],
       [-2.07563095,  1.48917752],
       [-2.44402884,  0.0476442 ],
       [-2.23284716,  0.22314807],
       [-2.33464048, -1.11532768],
       [-2.18432817, -0.46901356],
       [-2.1663101 ,  1.04369065],
       [-2.32613087,  0.13307834],
       [-2.2184509 , -0.72867617],
       [-2.6331007 , -0.96150673],
       [-2.1987406 ,  1.86005711],
       [-2.26221453,  2.68628449],
       [-2.2075877 ,  1.48360936],
       [-2.19034951,  0.48883832],
       [-1.898572  ,  1.40501879],
       [-2.34336905,  1.12784938],
       [-1.914323  ,  0.40885571],
       [-2.20701284,  0.92412143],
       [-2.7743447 ,  0.45834367],
       [-1.81866953,  0.08555853],
       [-2.22716331,  0.13725446],
       [-1.95184633, -0.62561859],
       [-2.05115137,  0.24216355],
       [-2.16857717,  0.52714953],
       [-2.13956345,

In [20]:
#recrearemos el mismo grafico 
result = []
for name in ('setosa', 'versicolor', 'virginica'):
    result = ir.Scatter(x=Y[y==name,0], y=Y[y==name,1],
                       mode = "markers", name=name, 
                       marker=ir.Marker(size=15, line=Line(color="rgba(225,225,225,0.2)", width=0.5),
                                       opacity = 0.75))
    results.append(result)
    
data = ir.Data(results)
layout=ir.Layout(xaxis= XAxis(title="CP1", showline=False),
                yaxis=YAxis(title="CP2", showline=False))

fig = ir.Figure(data=data, layout=layout)
fig.show()
    

AttributeError: 'Scatter' object has no attribute 'append'